### PR TITLE
Expand sparse.softmax zero nnz tests to cover cases of previously reported FPE.

### LINKED ISSUE #82107: CUDA sparse_softmax can trigger floating point exception
### 🐛 Describe the bug

```
>>> import torch
>>> torch.sparse.softmax(torch.sparse_coo_tensor([[]], [], (3,), device='cuda'),0)
Floating point exception (core dumped)
```

@cpuhrsch do you know who the right person for this is?

### Versions

master

cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95646

- Test cases with zero `nnz` added for `sparse.log_softmax`.
- Test cases with zero `nnz` for both `sparse.log_softmax` and
`torch.sparse_softmax` expanded to cover the backward pass.

These test additions prove resolution to #95371 and #82107.

Fixes #82107 #95371

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @bhosmer

### CODE REVIEW

**Review by cpuhrsch — APPROVED** [review]

File: test/test_sparse.py [hunk 1]
```diff
@@ -3481,6 +3481,25 @@ def test_softmax_zero_nnz(self, device, dtype):
         out = torch.sparse.softmax(t, 0)
         self.assertEqual(out.to_dense(), torch.zeros_like(t))
 
+        t = t.requires_grad_()
+
+        def fn(x):
+            return torch.sparse.softmax(x, 0).to_dense()
```
Comment (cpuhrsch) [inline test/test_sparse.py:3487]: Can we use straight gradcheck now that https://github.com/pytorch/pytorch/pull/95405 merged?

Comment (cpuhrsch) [reply test/test_sparse.py:3487]: Perhaps even enable OpInfo coverage?

Comment (amjames) [reply test/test_sparse.py:3509]: Gradcheck still does not support sparse outputs, so this wrapper is still needed.

Comment (cpuhrsch) [reply test/test_sparse.py:3509]: @pearu - Can we extend gradcheck to work for this case too?

Comment (pearu) [reply test/test_sparse.py:3509]: This works:
```python
>>> torch.autograd.gradcheck(lambda x: torch.sparse.softmax(x, 0).to_dense(), t, masked=True, check_sparse_nnz=True)
True
```

Here's a list of tasks for extending gradcheck:
1. Eliminate `check_sparse_nnz=` kw argument as redundant. This is an easy fix.
2. Allow op to return sparse output. I tried to disable the check --- there appear serious issues that need to be resolved to get sparse output working.

File: test/test_sparse.py [hunk 2]
```diff
@@ -3503,6 +3503,25 @@ def test_softmax_zero_nnz(self, device, dtype):
...
+        torch.autograd.gradcheck(fn, (t), check_sparse_nnz=True)
```
Comment (pearu) [inline test/test_sparse.py:3510]: Use:
```suggestion
        torch.autograd.gradcheck(fn, (t), check_sparse_nnz=True, masked=True)
```

File: test/test_sparse.py [hunk 3]
```diff
@@ -3503,6 +3503,25 @@ def test_softmax_zero_nnz(self, device, dtype):
...
+        t = torch.sparse_coo_tensor([[]], [], (3,), device=device, dtype=dtype)
+        out = torch.sparse.log_softmax(t, 0)
+        self.assertEqual(out.to_dense(), torch.zeros_like(t))
+
+        # check backward
+        t = t.requires_grad_()
+
+        def fn(x):
+            return torch.sparse.log_softmax(x, 0).to_dense()
+        torch.autograd.gradcheck(fn, (t), check_sparse_nnz=True)
```
Comment (pearu) [inline test/test_sparse.py:3523]: Use:
```suggestion
        torch.autograd.gradcheck(fn, (t), check_sparse_nnz=True, masked=True)
```

**Review by pearu — CHANGES_REQUESTED** [review]
`masked=True` is required when autograd is applied on operations from `torch.sparse` namespace.

Otherwise, LGTM!

File: test/test_sparse.py (hunk 1 above)
Comment (nikitaved) [reply test/test_sparse.py:3509]: Also nit: can we enable `float` dtype as well?

File: test/test_sparse.py [hunk 4]
```diff
@@ -3503,6 +3503,25 @@ def test_softmax_zero_nnz(self, device, dtype):
...
 
+        t = t.requires_grad_()
+
+        def fn(x):
+            return torch.sparse.softmax(x, 0).to_dense()
+        torch.autograd.gradcheck(fn, (t), check_sparse_nnz=True)
+
+    @dtypes(torch.float)
+    def test_log_softmax_zero_nnz(self, device, dtype):
+        t = torch.sparse_coo_tensor([[]], [], (3,), device=device, dtype=dtype)
```
Comment (nikitaved) [inline test/test_sparse.py:3514]: Are zero sizes implying zero nnz tested as well?

File: test/test_sparse.py (hunk 1 above)
Comment (pearu) [reply test/test_sparse.py:3509]: @nikitaved float32 is typically disabled for gradcheck tests because the numerical Jacobian result is often too inaccurate. To make the test work, tweaking of default eps/atol/rtol parameters is required.

File: test/test_sparse.py (hunk 4 above)
Comment (pearu) [reply test/test_sparse.py:3514]: There seems to be no issue with zero sizes:
```python
>>> t = torch.sparse_coo_tensor([[], []], [], (0, 3), dtype=torch.float64)
>>> t.requires_grad_()
tensor(indices=tensor([], size=(2, 0)),
       values=tensor([], size=(0,)),
       size=(0, 3), nnz=0, dtype=torch.float64, layout=torch.sparse_coo,
       requires_grad=True)
>>> torch.autograd.gradcheck(lambda x: torch.sparse.softmax(x, 0).to_dense(), t, masked=True, check_sparse_nnz=True)
True
```

File: test/test_sparse.py (hunk 1 above)
Comment (nikitaved) [reply test/test_sparse.py:3509]: These are trivial cases with no computation, so it should work for any dtype I reckon.

File: test/test_sparse.py (hunk 4 above)
Comment (nikitaved) [reply test/test_sparse.py:3514]: But is it tested? Looks like a great opportunity to cover degenerate cases.

File: test/test_sparse.py (hunk 1 above)
Comment (amjames) [reply test/test_sparse.py:3509]: Also curious if the "real" sparse softmax tests need to do all the stuff they are doing with computing and comparing the full Jacobian. Can't we just use gradcheck there as well? If we simplify that test then this just becomes an additional shape + nnz to pass as a test case.

Comment (pearu) [reply test/test_sparse.py]: I suggest keeping "the real sparse softmax tests" because it provides a closed form jacobian (as defined in `softmax_jacobian_analytic`) that would be the thrid independent approach of testing "sparse softmax" gradients together with the numerical and analytical jacobians from gradcheck.

**Review by pearu — APPROVED** [review]
LGTM! Thanks, @amjames!

**Review by nikitaved — APPROVED** [review]
👍
